### PR TITLE
Distance sensor

### PR DESCRIPTION
**Pull request by suspiciouscow**
added distance sesnor code

### CODE REVIEW

File: src/main/java/frc/team2412/robot/subsystems/IntakeSubsystem.java [hunk 1]
```diff
@@ -92,4 +95,8 @@ public boolean isSecured() { // Checks to see if the game piece is secured
 	}
 
 	public void rumble() {}
+
+	public double getDistance() {
+		return Math.pow(distanceSensor.getAverageVoltage(), -1.2045) * 27.726;
```
Comment (TAKBS2412) [inline src/main/java/frc/team2412/robot/subsystems/IntakeSubsystem.java:100]: Where did these numbers (and equation) come from?

Comment (suspiciouscow) [reply src/main/java/frc/team2412/robot/subsystems/IntakeSubsystem.java:100]: google apparently it converts it to cm

Comment (bobtseattle) [reply src/main/java/frc/team2412/robot/subsystems/IntakeSubsystem.java:101]: Please document that knowledge with a comment.

File: src/main/java/frc/team2412/robot/subsystems/IntakeSubsystem.java [hunk 2]
```diff
@@ -37,31 +38,34 @@ public static enum GamePieceType {
...
+		motor1.setIdleMode(IdleMode.kBrake);
+		motor2.setIdleMode(IdleMode.kBrake);
+		distanceSensor = new AnalogInput(INTAKE_DISTANCE_SENSOR);
 	}
 
 	// METHODS
 	public void setSpeed(double speed) {
-		motor.set(speed);
+		motor1.set(speed);
+		motor2.set(-speed);
```
Comment (bobtseattle) [inline src/main/java/frc/team2412/robot/subsystems/IntakeSubsystem.java:64]: If you find the need to start using motor commands in more methods, consider setting this motor to be inverted which can help make sure the sign is not forgotten.
Another option would be to setup motor2 to follow motor1.

File: src/main/java/frc/team2412/robot/subsystems/IntakeSubsystem.java [hunk 3]
```diff
@@ -37,31 +38,34 @@ public static enum GamePieceType {
...
 	public IntakeSubsystem() {
-		motor = new CANSparkMax(INTAKE_MOTOR_1, MotorType.kBrushless);
+		motor1 = new CANSparkMax(INTAKE_MOTOR_1, MotorType.kBrushless);
+		motor2 = new CANSparkMax(INTAKE_MOTOR_2, MotorType.kBrushless);
 		// colorSensor = new ColorSensorV3(Port.kOnboard); //to find I2C port
-		// distanceSensor = new ();
-		motor.setIdleMode(IdleMode.kBrake);
+		motor1.setIdleMode(IdleMode.kBrake);
+		motor2.setIdleMode(IdleMode.kBrake);
+		distanceSensor = new AnalogInput(INTAKE_DISTANCE_SENSOR);
```
Comment (bobtseattle) [inline src/main/java/frc/team2412/robot/subsystems/IntakeSubsystem.java:58]: Seems to be a white space issue here.  Make sure to apply spotless.

**Review by bobtseattle — COMMENTED** [review]
Some feedback

**Comment by eddiert1239** [comment]
In this latest commit, titled: "Color Sensor", we addressed the changes requested by our very own, bobtseattle. We also added a color sensor, with the port as kMXP, instead of kOnboard. We also added some logic in the isSecured() method, and detect the GamePieceType we are detecting from the color sensor in the detectType() method. In Controls.java, on the IntakeSetInCommand, we added an additional end condition, where if isSecured() returns true (either where the game object gets past a certain point on the distance sensor, or a cube is detected and it gets past an earlier point), the robot will stop intaking in.

File: src/main/java/frc/team2412/robot/subsystems/IntakeSubsystem.java [hunk 4]
```diff
@@ -77,19 +85,29 @@ public void intakeStop() {
...
+			return GamePieceType.CONE;
+		}
 		return GamePieceType.NONE;
 	}
 
-	public boolean isSecured() { // Checks to see if the game piece is secured
+	public boolean
+			isSecured() { // Checks to see if the game piece is secured, returns true if the motor should
+		// stop
+		if (getDistance() < 12) { // units in cm
```
Comment (bobtseattle) [inline src/main/java/frc/team2412/robot/subsystems/IntakeSubsystem.java:99]: It will be interesting to see how the intake reacts with cones not in expected orientations (tip during intake?)

File: src/main/java/frc/team2412/robot/Subsystems.java [hunk 5]
```diff
@@ -38,5 +41,8 @@ public Subsystems() {
 		if (INTAKE_ENABLED) {
 			intakeSubsystem = new IntakeSubsystem();
 		}
+		if (INTAKE_ENABLED) {
```
Comment (bobtseattle) [inline src/main/java/frc/team2412/robot/Subsystems.java:44]: Should this be LED_ENABLED?

**Review by bobtseattle — COMMENTED** [review]
Should there be a way for the drive team to turn off the leds?  If we are playing defense we may not want the human player to get ready to give us a game piece which distracts them from preparing to give a different piece to our alliance partner.

File: src/main/java/frc/team2412/robot/Subsystems.java (hunk 5 above)
Comment (TAKBS2412) [reply src/main/java/frc/team2412/robot/Subsystems.java:44]: Agreed, looks like the wrong flag to me.

File: src/main/java/frc/team2412/robot/subsystems/IntakeSubsystem.java [hunk 6]
```diff
@@ -77,19 +85,28 @@ public void intakeStop() {
...
 
 	public GamePieceType detectType() {
-		// if () {
-		//     return GamePieceType.CUBE;
-		// }
-		// else if () {
-		//     return GamePieceType.CONE;
-		// }
-
+		if (colorSensor.getColor().equals(INTAKE_CUBE_COLOR)) {
```
Comment (TAKBS2412) [inline src/main/java/frc/team2412/robot/subsystems/IntakeSubsystem.java:88]: Do we know if the color returned from the sensor will always be exactly equal to `INTAKE_CUBE_COLOR` or `INTAKE_CONE_COLOR`? I'm thinking about what would happen if the returned color changes due to lighting conditions, proximity, etc.

**Review by TAKBS2412 — CHANGES_REQUESTED** [review]

File: src/main/java/frc/team2412/robot/subsystems/IntakeSubsystem.java [hunk 7]
```diff
@@ -77,19 +85,28 @@ public void intakeStop() {
...
+		} else if (colorSensor.getColor().equals(INTAKE_CONE_COLOR)) {
+			return GamePieceType.CONE;
+		}
 		return GamePieceType.NONE;
 	}
 
-	public boolean isSecured() { // Checks to see if the game piece is secured
+	public boolean
+			isSecured() { // Checks to see if the game piece is secured, returns true if the motor should
+		// stop
```
Comment (KangarooKoala) [inline src/main/java/frc/team2412/robot/subsystems/IntakeSubsystem.java:98]: Nit: Move comment on a separate line to avoid weird line breaks (or even better, change it to a javadoc comment like this)
```java
/**
 * Checks to see if the game piece is secured
 * 
 * @return True if the motor should stop
 */
```

File: src/main/java/frc/team2412/robot/subsystems/IntakeSubsystem.java [hunk 8]
```diff
@@ -77,19 +85,28 @@ public void intakeStop() {
...
+		if (getDistance() < 12) { // units in cm
+			return true;
+		} else if (detectType() == GamePieceType.CUBE
+				&& getDistance() < 15) { // also arbitrary numbers right now
+			return true;
+		}
 		return false;
```
Comment (KangarooKoala) [inline src/main/java/frc/team2412/robot/subsystems/IntakeSubsystem.java:105]: Nit: Prefer using boolean logic (`return a || b && c`)

File: src/main/java/frc/team2412/robot/Subsystems.java (hunk 5 above)
Comment (eddiert1239) [reply src/main/java/frc/team2412/robot/Subsystems.java:44]: yeah oopsie

**Review by bobtseattle — APPROVED** [review]

**Review by TAKBS2412 — APPROVED** [review]

File: src/main/java/frc/team2412/robot/commands/intake/IntakeOutCommand.java [hunk 9]
```diff
@@ -0,0 +1 @@
+
```
Comment (KangarooKoala) [inline src/main/java/frc/team2412/robot/commands/intake/IntakeOutCommand.java:1]: Should this be put off to another PR? Is this supposed to be different from `IntakeSetOutCommand`?

File: src/main/java/frc/team2412/robot/commands/intake/IntakeSetOutCommand.java [hunk 10]
```diff
@@ -1,6 +1,7 @@
 package frc.team2412.robot.commands.intake;
 
 import static frc.team2412.robot.subsystems.IntakeSubsystem.IntakeConstants.INTAKE_OUT_SPEED;
+import static frc.team2412.robot.subsystems.LEDSubsystem.LEDConstants.*;
```
Comment (KangarooKoala) [inline src/main/java/frc/team2412/robot/commands/intake/IntakeSetOutCommand.java:4]: This doesn't do anything?

File: src/main/java/frc/team2412/robot/subsystems/IntakeSubsystem.java [hunk 11]
```diff
@@ -6,19 +6,25 @@
 import com.revrobotics.CANSparkMax;
 import com.revrobotics.CANSparkMax.IdleMode;
 import com.revrobotics.CANSparkMaxLowLevel.MotorType;
+import com.revrobotics.ColorSensorV3;
+import edu.wpi.first.wpilibj.AnalogInput;
+import edu.wpi.first.wpilibj.I2C.Port;
 import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import java.awt.Color;
```
Comment (KangarooKoala) [inline src/main/java/frc/team2412/robot/subsystems/IntakeSubsystem.java:13]: Probably should use [`edu.wpi.first.wpilibj.util.Color`](https://github.wpilib.org/allwpilib/docs/release/java/edu/wpi/first/wpilibj/util/Color.html) instead of [`java.awt.Color`](https://docs.oracle.com/en/java/javase/11/docs/api/java.desktop/java/awt/Color.html), though in our use case (storing RGB values) they work identically.

**Review by KangarooKoala — COMMENTED** [review]
Could merge anyways but there's some questions and code quality stuff that would preferably be addressed (but you could mark them as resolved to merge faster).

File: src/main/java/frc/team2412/robot/subsystems/IntakeSubsystem.java (hunk 11 above)
Comment (eddiert1239) [reply src/main/java/frc/team2412/robot/subsystems/IntakeSubsystem.java:13]: I was trying to use the wpilib one, but I couldn't find methods returning the individual r g and b values, so I resorted to this one since it worked

Comment (KangarooKoala) [reply src/main/java/frc/team2412/robot/subsystems/IntakeSubsystem.java:13]: Hm, looks like [`Color8Bit`](https://github.wpilib.org/allwpilib/docs/release/java/edu/wpi/first/wpilibj/util/Color8Bit.html) would be usable (also it can be passed into `AddressableLEDBuffer` if you feel like using it there too), but it should work either way.

Comment (TAKBS2412) [reply src/main/java/frc/team2412/robot/subsystems/IntakeSubsystem.java:13]: For the wpilib one you could use the `red`, `green`, and `blue` fields to get the individual values (it's pretty far down on the documentation page) but as Joseph mentioned, it's not too important.

File: src/main/java/frc/team2412/robot/commands/intake/IntakeOutCommand.java (hunk 9 above)
Comment (suspiciouscow) [reply src/main/java/frc/team2412/robot/commands/intake/IntakeOutCommand.java:2]: we need both we need IntakeOut for IntakeSetOut

File: src/main/java/frc/team2412/robot/commands/intake/IntakeSetOutCommand.java (hunk 10 above)
Comment (suspiciouscow) [reply src/main/java/frc/team2412/robot/commands/intake/IntakeSetOutCommand.java]: removed

File: src/main/java/frc/team2412/robot/subsystems/IntakeSubsystem.java (hunk 11 above)
Comment (suspiciouscow) [reply src/main/java/frc/team2412/robot/subsystems/IntakeSubsystem.java:13]: Can be fixed in a future pr, we'll circle back 😉😏

File: src/main/java/frc/team2412/robot/subsystems/IntakeSubsystem.java [hunk 12]
```diff
@@ -6,19 +6,26 @@
...
-		public static final double INTAKE_OUT_SPEED = 0.3;
+		public static final double INTAKE_HOLD_SPEED = 0.1;
+		public static final double INTAKE_IN_SPEED = 0.8;
+		public static final double INTAKE_OUT_SPEED = -0.8;
 
 		public static final double INTAKE_CUBE_DISTANCE = 0;
 		public static final double INTAKE_CONE_DISTANCE = 0;
-		public static final double INTAKE_CUBE_COLOR = 0;
-		public static final double INTAKE_CONE_COLOR = 141;
+		public static final Color INTAKE_CUBE_COLOR = new Color(145, 48, 255);
```
Comment (bobtseattle) [inline src/main/java/frc/team2412/robot/subsystems/IntakeSubsystem.java:25]: The colors are set in two different places.  It would be good to use a constants so they do not accidentally differ.

File: src/main/java/frc/team2412/robot/subsystems/LEDSubsystem.java [hunk 13]
```diff
@@ -1,21 +1,39 @@
...
 import edu.wpi.first.wpilibj.AddressableLEDBuffer;
 import edu.wpi.first.wpilibj2.command.SubsystemBase;
 
 public class LEDSubsystem extends SubsystemBase {
-	private final int BUFFER_LENGTH = 72;
+	// CONSTANTS
+	public static class LEDConstants {
+		public static final int BUFFER_LENGTH = 72;
+
+		// private final RGB PURPLE_RGB = 145, 48, 255;
```
Comment (bobtseattle) [inline src/main/java/frc/team2412/robot/subsystems/LEDSubsystem.java:15]: The other place the colors are set.

**Review by bobtseattle — CHANGES_REQUESTED** [review]
Almost there.

File: src/main/java/frc/team2412/robot/subsystems/LEDSubsystem.java (hunk 13 above)
Comment (suspiciouscow) [reply src/main/java/frc/team2412/robot/subsystems/LEDSubsystem.java:16]: everything be in intake game piece type enum now

File: src/main/java/frc/team2412/robot/subsystems/IntakeSubsystem.java [hunk 14]
```diff
@@ -77,19 +85,38 @@ public void intakeStop() {
...
-		// if () {
-		//     return GamePieceType.CUBE;
-		// }
-		// else if () {
-		//     return GamePieceType.CONE;
-		// }
-
+		if (colorSensorEquals(CUBE.color)) {
+			return GamePieceType.CUBE;
+		} else if (colorSensorEquals(CUBE.color)) {
```
Comment (bobtseattle) [inline src/main/java/frc/team2412/robot/subsystems/IntakeSubsystem.java:90]: CONE?

**Review by bobtseattle — CHANGES_REQUESTED** [review]
Fix cut and paster error.

**Review by bobtseattle — APPROVED** [review]
